packages/loopring_v3/contracts/impl/libtransactions/OwnerChangeTransaction.sol

```diff
@@ -0,0 +1,189 @@
+            ctx.DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encode(
+                    OWNERCHANGE_TYPEHASH,
+                    accountTransfer.owner,
+                    accountTransfer.accountID,
+                    accountTransfer.feeTokenID,
+                    accountTransfer.fee,
+                    accountTransfer.newOwner,
+                    accountTransfer.nonce,
+                    auxData.walletAddress,
+                    auxData.walletDataHash,
+                    keccak256(auxData.walletCalldata)
+                )
+            )
+        );
+
+        // Verify the signature if one is provided, otherwise fall back to an approved tx
+        if (auxData.signatureNewOwner.length > 0) {
+            require(txHash.verifySignature(accountTransfer.newOwner, auxData.signatureNewOwner), "INVALID_SIGNATURE");
+        } else {
+            require(S.approvedTx[accountTransfer.newOwner][txHash], "TX_NOT_APPROVED");
+            S.approvedTx[accountTransfer.newOwner][txHash] = false;
+        }
+
+        if (auxData.walletAddress == address(0)) {
+            // We allow the owner of the account to change when authorized by the current owner.
+            // Verify the signature if one is provided, otherwise fall back to an approved tx
+            if (auxData.signatureOldOwner.length > 0) {
+                require(txHash.verifySignature(accountTransfer.owner, auxData.signatureOldOwner), "INVALID_SIGNATURE");
+            } else {
+                require(S.approvedTx[accountTransfer.owner][txHash], "TX_NOT_APPROVED");
+                S.approvedTx[accountTransfer.owner][txHash] = false;
+            }
+        } else {
+            // If the account has a wallet, use it to recover the account
+            require(accountTransfer.walletHash != 0, "ACCOUNT_HAS_NO_WALLET");
+
+            // Calculate the wallet hash
+            bytes32 walletHash = EIP712.hashPacked(
+                S.DOMAIN_SEPARATOR,
+                keccak256(
+                    abi.encode(
+                        WALLET_TYPEHASH,
+                        auxData.walletAddress,
+                        auxData.walletDataHash
+                    )
+                )
+            );
+            // Hashes are stored using only 253 bits so the value fits inside a SNARK field element.
+            require((uint(walletHash) >> 3) == uint(accountTransfer.walletHash), "INVALID_WALLET_HASH");
+
+            // Check that the calldata contains the correct inputs for data coming from layer 2.
+            // This data is also padded with zeros to 32 bytes (at the MSB) inside the calldata.
+            // parameter 0: accountID
+            // parameter 1: nonce
+            // parameter 2: oldOwner
+            // parameter 3: newOwner
+            // parameter 4: walletDataHash
+            uint offset = 4;
+            require(auxData.walletCalldata.toUint24(29 + offset) == accountTransfer.accountID, "INVALID_WALLET_CALLDATA");
```

Comment:
instead of putting the duplicate data inside auxData check for consistency, can we simply use a new Struct to wrap them, then simply concatenate it with 'welletCalldata':

```
struct CommonParams {
   ...
}

CommonParam params = abi.encodePacked(CommonParam(...));

(bool success, bytes memory result) =
    auxData.walletAddress.staticcall(
        abi.encodePacked(params, auxData.walletCalldata)
    );
```

In StateWallet, we need to define a matching CommonParams struct and always use it as the first param for external functions.

Reply:
`abi.encodePacked(params, auxData.walletCalldata)`

This won't encode the data correctly for a function call, but something like this would certainly be possible. I'm not sure if it will end up being easier/better or not, would have to think about it some more.

Reply:
Please give it a try, so we have can reduce the calldata size as much as possible.

Reply:
> Please give it a try, so we have can reduce the calldata size as much as possible

I don't really care for the calldata size here to be honest. This operation will be expensive for other reasons (signature checks etc...). I only care about making this as simple as possible.